torch/utils/_pytree.py

```diff
@@ -26,194 +28,124 @@
 To improve the performance we can move parts of the implementation to C++.
 """
 
+PROTOCOL_VERSION = 1
+
+Context = Any
+PyTree = Any
+FlattenFunc = Callable[[PyTree], Tuple[List, Context]]
+UnflattenFunc = Callable[[List, Context], PyTree]
+DumpableContext = Any  # Any json dumpable text
+GetStateFn = Callable[[Context], DumpableContext]
+SetStateFn = Callable[[DumpableContext], Context]
+
 # A NodeDef holds two callables:
 # - flatten_fn should take the collection and return a flat list of values.
 #   It can also return some context that is used in reconstructing the
 #   collection.
 # - unflatten_fn should take a flat list of values and some context
 #   (returned by flatten_fn). It returns the collection by reconstructing
 #   it from the list and the context.
-# - to_str_fn takes a TreeSpec with the specific type and a list of its children
-#   TreeSpecs already converted to strings, and returns a string representation
-#   of this TreeSpec
-# - maybe_from_str_fn takes in a string and if this string represents a TreeSpec
-#   of this type, returns the type, the context, and a string representation of
-#   its children specs. Otherwise it returns None.
-Context = Any
-PyTree = Any
-FlattenFunc = Callable[[PyTree], Tuple[List, Context]]
-UnflattenFunc = Callable[[List, Context], PyTree]
-ToStrFunc = Callable[["TreeSpec", List[str]], str]
-MaybeFromStrFunc = Callable[[str], Optional[Tuple[Any, Context, str]]]
-
 class NodeDef(NamedTuple):
     type: Type[Any]
     flatten_fn: FlattenFunc
     unflatten_fn: UnflattenFunc
-    to_str_fn: ToStrFunc
-    maybe_from_str_fn: MaybeFromStrFunc
 
-SUPPORTED_NODES: Dict[Type[Any], NodeDef] = {}
+SUPPORTED_NODES: Dict[Union[Type[Any], str], NodeDef] = {}
+
+# SerializeNodeDef holds two callables:
+# - serialize_fn takes a TreeSpec, and returns a serialized string format of the
+#   context, and the version number
+# - deserialize_fn takes in a string representation of the context, and the
+#   version, and returns the deserialized context
+class SerializeNodeDef(NamedTuple):
+    typ: Type[Any]
+    serialized_type: str
+    getstate: Optional[GetStateFn]
+    setstate: Optional[SetStateFn]
```

Comment:
Thing about it a little more, it's kind of weird that we need to ask users "how do you want to serialize your context". Because what it seems like we really want to ask the user is "how do you want to serialize this unserializable object I found in your context?"

Maybe this is fine